blog/modules/ROOT/pages/using-types.adoc

```diff
@@ -0,0 +1,256 @@
+<1> `name` cannot be empty or a blank String.
+<2> `email` must contain a `@` symbol.
+<3> A Swift file must have the correct extension.
+<4> Declaring some common possibilities before the generic type
+    is a useful pattern that results in better tooling support in IDEs.
+
+There are plenty of ways to make our types more fine-grained than just `String`,
+but strings are not the only types we can refine in Pkl.
+
+== Constraining numbers
+
+Pkl comes with some useful typealiases for numbers like `Int16`, `UInt8`, etc.
+
+[source,pkl]
+----
+serverPort: UInt16 // <1>
+----
+<1> Will only allow values between 0 and 65535, inclusive.
+
+Builtin typealiases are not the only way to handle such cases, though.
+Numbers can be constrained to any value.
+
+[source,pkl]
+----
+volume: Float(isBetween(0.0, 11.0)) // <1>
+
+multiplier: Int(isEven) // <2>
+----
+<1> `volume` can only be in the range between 0 and 11.
+<2> `multiplier` cannot be odd.
+
+== Beyond builtins
+
+Up to now, we only took a look at builtin functions, mostly functions of the type itself.
+We also limited our types to a single constraint. Pkl allows, though, for arbitrary functions as constraints,
+and we can provide multiple ones:
+
+[source,pkl]
+----
+local isPrime = (n: Int) ->
+  n > 1 && IntSeq(2, n - 1).toList().every((x) -> n % x != 0)
+
+hash: Int(isPrime) // <1>
+
+region: String(startsWith("us-"), length < 10) // <2>
+
+name: String(this == capitalize()) // <3>
+
+building: String(isEmpty || toInt() <= 100) // <3>
+
+grade: Int(isBetween(0, 10))|String(matches(Regex(#"[A-F][+-]?"#))) // <4>
+----
+<1> User defined constraints.
+<2> Multiple constraints.
+<3> Arbitrary boolean expressions.
+<4> Multiple possibilities with different constraints.
+
+== Better errors
+
+One downside of using arbitrary expressions as constraints is that, sometimes, errors messages
+can be a bit hard to understand:
+
+[source,pkl]
+----
+ssn: String(matches(Regex(#"\d{3}-\d{2}-\d{4}"#))) = "123-123-123"
+----
+
+[source,text]
+----
+–– Pkl Error ––
+Type constraint `matches(Regex(#"\d{3}-\d{2}-\d{4}"#))` violated.
+Value: "123-123-123"
+
+1 | ssn: String(matches(Regex(#"\d{3}-\d{2}-\d{4}"#))) = "123-123-123"
+                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+...
+----
+
+We can do better than that.
+
+One trick to improve error messages is to create named lambdas:
+
+[source,pkl]
+----
+local isValidSocialSecurityNumber = (str: String) ->
+  str.matches(Regex(#"\d{3}-\d{2}-\d{4}"#))
+
+ssn: String(isValidSocialSecurityNumber) = "123-123-123"
+----
+
+[source,text]
+----
+–– Pkl Error ––
+Type constraint `isValidSocialSecurityNumber` violated.
+Value: "123-123-123"
+
+4 | ssn: String(isValidSocialSecurityNumber) = "123-123-123"
+                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+...
+----
+
+That's already better. Users now have a hint of what they did wrong.
+
+=== Providing custom error messages
+
+We just saw how to improve error messages with lambdas, but Pkl can
+do even better than that.
+
```

Comment:
```suggestion
Sometimes, it can be useful to provide your own custom error messages.
Pkl doesn't provide a built-in way to do this, but one pattern to hack around this limitation is to use `throw`.

WARNING: This approach can be a footgun. See <<dangers-of-throw>> for details.
```